services/billing.py

```diff
@@ -155,18 +157,75 @@ def delete_subscription(self, owner: Owner):
             stripe.SubscriptionSchedule.release(subscription_schedule_id)
 
-        stripe.Subscription.modify(
-            owner.stripe_subscription_id,
-            cancel_at_period_end=True,
-            proration_behavior="none",
+        # we give an auto-refund grace period of 24 hours for a monthly subscription or 72 hours for a yearly subscription
+        current_subscription_datetime = datetime.fromtimestamp(
+            subscription["current_period_start"]
+        )
+        difference_from_now = datetime.now() - current_subscription_datetime
+
+        subscription_plan_interval = getattr(
+            getattr(subscription, "plan", None), "interval", None
         )
+        within_refund_grace_period = (
+            subscription_plan_interval == "month" and difference_from_now.days < 1
+        ) or (subscription_plan_interval == "year" and difference_from_now.days < 3)
+
+        customer = stripe.Customer.retrieve(owner.stripe_customer_id)
+        # we are giving customers 2 autorefund instances
+        autorefunds_remaining = int(
+            customer["metadata"].get("autorefunds_remaining", "2")
+        )
+
+        if within_refund_grace_period and autorefunds_remaining > 0:
+            stripe.Subscription.cancel(owner.stripe_subscription_id)
+
+            invoices_list = stripe.Invoice.list(
+                subscription=owner.stripe_subscription_id, status="paid"
+            )
+            created_refund = False
+            # there could be multiple invoices that need to be refunded such as if the user increased seats within the grace period
+            for invoice in invoices_list["data"]:
+                start_of_last_period = (
+                    current_subscription_datetime - relativedelta(months=1)
+                    if subscription_plan_interval == "month"
+                    else current_subscription_datetime - relativedelta(years=1)
+                )
+
+                # refund if all of the following are true: the invoice has a charge, it has been fully paid,
+                # the creation time was before the start of the current subscription's start, and the creation time was after the start of the last period
+                invoice_created_datetime = datetime.fromtimestamp(invoice["created"])
+                if (
+                    invoice["charge"] is not None
+                    and invoice["amount_remaining"] == 0
+                    and invoice_created_datetime < current_subscription_datetime
+                    and invoice_created_datetime >= start_of_last_period
+                ):
+                    stripe.Refund.create(invoice["charge"])
+                    created_refund = True
+            if created_refund == True:
+                stripe.Customer.modify(
+                    owner.stripe_customer_id,
+                    balance=0,
+                    metadata={"autorefunds_remaining": str(autorefunds_remaining - 1)},
+                )
+                log.info(
```

Comment:
any other log info that might be useful here? Maybe customer id and sub id off the top of my head, we can add them with the extra=dict() param